### DISCUSSION TITLE
Running app error when executed with `/lib/ld-musl-x86_64.so.1`.

### DISCUSSION THREAD

**Discussion by GamePlayer-8** [pyinstaller/pyinstaller#7551] [Help] [answered]
Heyo,

I have a little problem with running compiled a compiled PyInstaller file inside AppImage, built on Alpine Linux.
The error what I'm getting says:
```shell
[22992] Cannot open PyInstaller archive from executable (/tmp/.mount_ticloctzbPnm/lib/ld-musl-x86_64.so.1) or external archive (/tmp/.mount_ticloctzbPnm/lib/ld-musl-x86_64.so.1.pkg)
```

Is there any way to tell PyInstaller to look for a second argument (`$1`) instead of first one (`$0`), which is `lib/ld-musl-x86_64.so.1`?

The `AppRun` script looks like so:
```shell
#!/bin/sh
TICLOCK_RUNPATH="$(dirname "$(readlink -f "${0}")")"
TICLOCK_EXEC="${TICLOCK_RUNPATH}"/usr/bin/ticlock
export LD_LIBRARY_PATH="${TICLOCK_RUNPATH}"/lib:"${TICLOCK_RUNPATH}"/lib64:$LD_LIBRARY_PATH
export LIBRARY_PATH="${TICLOCK_RUNPATH}"/lib:"${TICLOCK_RUNPATH}"/lib64:"${TICLOCK_RUNPATH}"/usr/lib:"${TICLOCK_RUNPATH}"/usr/lib64:$LIBRARY_PATH
export PATH="${TICLOCK_RUNPATH}/usr/bin/:${TICLOCK_RUNPATH}/usr/sbin/:${TICLOCK_RUNPATH}/usr/games/:${TICLOCK_RUNPATH}/bin/:${TICLOCK_RUNPATH}/sbin/${PATH:+:$PATH}"
exec "${TICLOCK_RUNPATH}"/lib/ld-musl-x86_64.so.1 "${TICLOCK_EXEC}" "$@"
```
where `ticlock` is a compiled PyInstaller file.

**Comment by rokm** [comment]
> Is there any way to tell PyInstaller to look for a second argument (`$1`) instead of first one (`$0`), which is `lib/ld-musl-x86_64.so.1`?

That's not really what is happening - the `argv[0]` is in fact the executable passed via `${TICLOCK_EXEC}`, as one would expect. But what happens is that PyInstaller's bootloader tries to look up its actual name via `/proc/self/exe`, and that causes it to end up with the path to `ld-musl-x86_64.so.1`:

https://github.com/pyinstaller/pyinstaller/blob/6fc2cc0cd9ec57edaf113ce1166abb02bc64558a/bootloader/src/pyi_path.c#L297-L303

So if you want to run the executable via `ld-musl-x86_64.so.1`, you'll need to disable that block in bootloader's code, rebuild the bootloader, and then rebuild your program with updated bootloader.

The `/proc/self/exe` resolution was re-added in #5232 in response to several issue reports, so I think it will stay there by default. But I'm not opposed to allowing it to be disabled via environment variable (e.g., PYINSTALLER_EXECUTABLE_NAME_LOOKUP=0)  to cover scenarios such as yours.

**Comment by GamePlayer-8** [comment]
So at the moment I need to recompile PyInstaller?

**Reply by rokm** [reply]
Yes, you need to check out PyInstaller's source, modify and recompile the bootloader, and then install and use the modified PyInstaller version.

**Comment by rokm** [comment, marked as answer]
This should be fixed in next release (#7556); the bootloader will try to detect if `/proc/self/exe` points to `ld.so`, and fall back to manual executable path resolution.

**Reply by GamePlayer-8** [reply]
Thank you!
